plugins/hls-rename-plugin/src/Ide/Plugin/Rename.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+
+module Ide.Plugin.Rename (descriptor) where
+
+import           Data.Function
+import           Data.HashMap.Internal               hiding (map)
+import           Data.List
+import qualified Data.Text                           as T
+import           Development.IDE                     hiding (pluginHandlers)
+import           Development.IDE.LSP.HoverDefinition
+import           Ide.Types
+import           Language.LSP.Server
+import           Language.LSP.Types
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor pluginId = (defaultPluginDescriptor pluginId) {
+    pluginHandlers = mkPluginHandler STextDocumentRename renameProvider
+}
+
+renameProvider :: PluginMethodHandler IdeState TextDocumentRename
+renameProvider state _pluginId (RenameParams tdi pos _progToken name) = do
+    locs <- getTextEdits state tdi pos name
+    return $ Right (WorkspaceEdit {
+        _changes = Just (fromList locs),
+        _documentChanges = Nothing,
+        _changeAnnotations = Nothing
+    })
+
+getTextEdits :: IdeState
+    -> TextDocumentIdentifier
+    -> Position
+    -> T.Text
+    -> LspT c IO [(Uri, List TextEdit)]
+getTextEdits state tdi pos name = do
+    mbLocs <- references state $ ReferenceParams tdi pos Nothing Nothing (ReferenceContext False)
```

Comment:
In a multi-project cradle, `references` only returns source locations for the projects that have been indexed. If not all the projects have been indexed, this will lead to incomplete renames. 

How can we address this? 

One way might be to grep the workspace for occurrences of the identifier and `uses GetHieAst` them or similar before calling `references` to ensure that they get indexed.

Reply:
My initial idea was to use the `refMap` from the `HieAstResult` for the files of interest and reference from the DB otherwise. I've done this in my most recent push. Would this be sufficient?

Reply:
I suspect it won't, because it doesn't address my point - the DB gets populated when files are type checked, and in a multi-project cradle this only happens for the projects that are loaded.

Take the `GhcSessionDeps` data constructor, defined in `Development.IDE.Core.RuleTypes` and used in a bunch of HLS plugins. Can you rename all the occurrences successfully? If yes, then close your editor, wipe `.cache/ghcide`, restart your editor and try the rename again - does it rename all the occurrences this time?

Reply:
I think this is a limitation we will just have to live with. I've been asking @fendor for a long time on a way to enumerate all the components in a cradle.

Reply:
> One way might be to grep the workspace for occurrences of the identifier and uses GetHieAst them or similar before calling references to ensure that they get indexed.

I think this is not a very nice solution at all. For example, naively implemented it will raise all sorts of when we try to `GetHieAST` on files which can't be loaded due to not having a proper crade.

The real solution would be to work on hie-bios and define an interface to enumerate all the components and subprojects in a particular cradle.

Reply:
> I think this is a limitation we will just have to live with. I've been asking @fendor for a long time on a way to enumerate all the components in a cradle.

A feature that only does X% of the job, X% of the time is bound to generate a lot of support traffic, and I wouldn't want to be on the maintainer side of it.

> The real solution would be to work on hie-bios and define an interface to enumerate all the components and subprojects in a particular cradle.

Can you and @fendor draft the work needed here? Maybe @OliverMadine could take a look at it as part of the GSoc project.

Reply:
I’d be happy to do work on this. A draft of the work would be much appreciated.